text/3476-expose-fn-type.md

```diff
@@ -0,0 +1,164 @@
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As described in the **Guide-level explanation**, with the syntax `fn <fn_path>`, we can reference the type behind the named function.
+
+When the function is for example in a different mod, it should be referenced by its path
+```rust
+mod sub {
+    fn sub_mod_fn() { .. }
+}
+impl fn sub::sub_mod_fn {
+    /* ... */
+}
+```
+
+It should be also possible to get the type of functions inside impl blocks:
+
+```rust
+struct MyStruct;
+impl MyStruct {
+    fn new() -> Self { Self }
+}
+impl fn MyStruct::new {
+    /* ... */
+}
+```
+
+When a function has generics, they will be handled as follows, just like we know it from normal types
+```rust
+fn send<T: Send>(val: T) {}
+impl<T: Send> ParcelStation for fn send<T> {
+    /* ... */
+}
+```
+
+When we have an implicit generic, they will be appended in order to the generic list:
+```rust
+fn implicit_generic(val: impl Clone) -> impl ToString {}
+impl<T: Send, U: ToString> for fn implicit_generic<T, U> {
+    /* ... */
+}
+```
+
+Just as structs and enums have the possibility to derive traits to automatically generate code, function type do too
+
+```rust
+#[derive(DbgSignature)]
+fn signature_test(val: i32) -> bool {
+    /* ... */
+}
+
+// Expands to
+
+fn signature_test(val: i32) -> bool {
+    /* ... */
+}
+impl DbgSignature for fn signature_test {
+    fn dbg_signature() -> &'static str {
+        "fn signature_test(val: i32) -> bool"
+    }
+}
+```
+
+Other than that, it should behave like every other type does.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- When introducing the derive feature, it could lead to parsing problems with proc macros having an older `syn` crate version.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The type behind functions already exists, we just need to expose it to the user.
+The hard part would be allowing derives, because that may break some things.
+
+# Prior art
+[prior-art]: #prior-art
+
+i dont know any
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Is the syntax good? It could create confusion between a function pointer.
+- What about closures? They don't even have names so targetting them would be quite difficult. I wouldn't want to use the compiler generated mess of a name like `[closure@src/main.rs:13:18: 13:20]`. It would also contain line numbers which would be changing quite often so thats not ideal.
+- I provided a possible solution for a `fn implicit_generic(val: impl Clone) -> impl ToString` function, but because we currently don't have a defined syntax for those generics in types, thus we can't use `impl Trait` as types for fields in structs, we should think about this more, maybe don't implement exposed types of function for such `fn`s and wait for another RFC?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- Also expose the type of closures
```

Comment:
Consider including consistency of display names as well. Currently:

- `std::any::type_name_of_val(&drop::<()>)` evaluates to `"core::mem::drop<()>"`
- When a specific function pointer appears in a compiler error, it looks like this:

```
5 |     let () = drop::<()>;
  |         ^^   ---------- this expression has type `fn(()) {std::mem::drop::<()>}`
  |         |
  |         expected fn item, found `()`
```

Choices of display include "fn item" (as opposed to "fn pointer" if it is cast to `fn(())` first), the function path and a mix of function pointer + `{path}`. It might be more consistent if the syntax eventually adopted in this RFC is consistent with the syntax in compiler diagnostics.

Reply:
this is a fair point, but i feel like that this will just make more boilerplate. we already specified the function and i dont want to make my impl even longer... on the other side: WAIT A MINUTE! this could solve our "what to do with `impl Trait` type param" problem...
So that we write
```rust
fn my_fn(x: impl ToString) -> String { x.to_string() }
impl<X: ToString> fn(X) my_fn {
   // ...
}
```
Ok that would be really cool.
So would you agree that boilerplate is okay here? @SOF3 

"Problem" is that this RFC tends to turn into a more generic `type_of` thing where this syntax wouldn't be possible anymore.. but type_of is another story so what you requested might be the solution i needed

Reply:
So i am currently rewriting and for now I will do this partially but use `fn_name` rather than `{fn_name}` because i think this could collide when we want to do something like this
```rust
impl MyTrait for fn() -> bool {

}
```
I don't know if something like this is planned or if this is already marked as "no" (if so please say it to me so ill change it in the RFC), but if it isn't, then the compiler may not be able to differentiate between `fn() {name} {block}` and `fn() {block}`

Reply:
> this is a fair point, but i feel like that this will just make more boilerplate. we already specified the function and i dont want to make my impl even longer... on the other side: WAIT A MINUTE! this could solve our "what to do with `impl Trait` type param" problem... So that we write
> 
> ```rust
> fn my_fn(x: impl ToString) -> String { x.to_string() }
> impl<X: ToString> fn(X) my_fn {
>    // ...
> }
> ```
> 
> Ok that would be really cool. So would you agree that boilerplate is okay here? @SOF3
> 
> "Problem" is that this RFC tends to turn into a more generic `type_of` thing where this syntax wouldn't be possible anymore.. but type_of is another story so what you requested might be the solution i needed

not a fan of this. two type expressions joined together without a delimiter is most likely not acceptable to the compiler team, considering we can't do the same with decl macro inputs either.

Reply:
> 

what about just `fn {fn_name}`?

Reply:
What do you mean by joined together?

Reply:
if it has a return type, the return expression will immediately be followed by the function path, which might lead to parsing ambiguities

Reply:
I don't think that's a problem.

Reply:
So what about moving the name front? `fn my_func` and `fn my_func(prms) -> ret`. But that would not match with the error syntax u sent

Reply:
actually my point was that we should make them consistent, but we could change the diagnostic display instead of the syntax

Reply:
So is the syntax that I currently have in the RFC ok?

Reply:
I added a `ToDo's` section at the end where i proposed a syntax change but I'm not sure if it is that well with the generic in the for "structure"? 😅